zebra-chain/src/orchard/tree.rs

```diff
@@ -154,70 +153,104 @@ impl ZcashDeserialize for Root {
     }
 }
 
-/// Orchard Note Commitment Tree
+/// Orchard Incremental Note Commitment Tree
 #[derive(Clone, Debug, Default, Eq, PartialEq)]
 struct NoteCommitmentTree {
-    /// The root node of the tree (often used as an anchor).
-    root: Root,
-    /// The height of the tree (maximum height for Orchard is 32).
-    height: u8,
-    /// The number of leaves (note commitments) in this tree.
-    count: u32,
+    left: Option<pallas::Base>,
+    right: Option<pallas::Base>,
+    parents: Vec<Option<pallas::Base>>,
 }
 
-impl From<Vec<NoteCommitment>> for NoteCommitmentTree {
-    fn from(_values: Vec<NoteCommitment>) -> Self {
-        unimplemented!();
-    }
-}
-
-impl From<Vec<pallas::Base>> for NoteCommitmentTree {
-    fn from(values: Vec<pallas::Base>) -> Self {
-        if values.is_empty() {
-            return NoteCommitmentTree {
-                root: Root::default(),
-                height: 0,
-                count: 0,
-            };
+impl NoteCommitmentTree {
+    /// Adds a note commitment x-coordinate to the tree.
+    ///
+    /// The leaves of the tree are actually a base field element, the
+    /// x-coordinate of the commitment, the data that is actually stored on the
+    /// chain and input into the proof.
+    ///
+    /// Returns an error if the tree is full.
+    #[allow(clippy::identity_op)]
+    pub fn append(&mut self, cm_x: pallas::Base) -> Result<(), ()> {
```

Comment:
This code was copied from `librustzcash`, so we need to document that.

It also means we want to avoid modifying the code as much as possible.
So I'll go resolve the rest of my suggestions.

```suggestion
    /// Returns an error if the tree is full.
    //
    // This code is based on https://github.com/zcash/librustzcash/blob/4c4d226f404aef64df6a77caa289b30cc875ae4f/zcash_primitives/src/merkle_tree.rs#L121
    #[allow(clippy::identity_op)]
    pub fn append(&mut self, cm_x: pallas::Base) -> Result<(), ()> {
```